README.md
### Foreword
F*** spammers. If you're one of them, or you operate shady marketing campaigns, please get lost.


For recent changes, refer to the [Change Log](https://github.com/venomous0x/WhatsAPI/blob/master/CHANGELOG.md)

# WhatsAPI

Interface to WhatsApp Messenger

----------


### What is WhatsApp?
According to [the company](http://www.whatsapp.com/): 

> “WhatsApp Messenger is a cross-platform mobile messenger that replaces SMS and works through the existing internet data plan of your device. WhatsApp is available for iPhone, BlackBerry, Android, Windows Phone, Nokia Symbian60 & S40 phones. Because WhatsApp Messenger uses the same internet data plan that you use for email and web browsing, there is no cost to message and stay in touch with your friends.”

Late 2011 numbers: 1 billion messages per day, ~20 million users.

### Modified XMPP
WhatsApp uses some sort of customized XMPP server, named internally as FunXMPP, which is basically some extended proprietary version.

### Login procedure
Much like XMPP, WhatsApp uses JID (jabber id) and password to successfully login to the service. The password is hashed and happened to be an MD5’d, reversed-version of the mobile’s IMEI (International Mobile Equipment Identity) or equivalent unique ID, stored in servers upon account creation and used transparently everytime the client connects the server.


The JID is a concatenation between your country’s code and mobile number.

Initial login uses Digest Access Authentication.

### Message sending
Messages are basically sent as TCP packets, following WhatsApp’s own format (unlike what’s defined in XMPP RFCs).


Despite the usage of SSL-like communication, messages are being sent in plain-text format.

### Multimedia Message sending
Photos, Videos and Audio files shared with WhatsApp contacts are HTTP-uploaded to a server before being sent to the recipient(s) along with Base64 thumbnail of media file (if applicable) along with the generated HTTP link as the message body.

# FAQ


- **What’s with the hex chars floating all over the code?**

	Mostly WhatsApp’s proprietary control chars/commands, or formatted data according to their server’s specifications, stored in predefined dictionaries within the clients.

- **What’s your future development plans?**

	We don’t have any.

- **Would it run over the web?**

	We’ve tested a slightly-modified version on top of Tornado Web Server and worked like a charm, however, building a chat client is a bit tricky, do your research.

- **Can I receive chats?**

	Indeed, using the same socket-receiving mechanism. But you have to parse the incoming data. Parsing functions aren’t included in this release, maybe in the next one?

- **I think the code is messy.**

	It’s working.

- **How can I obtain my password?**

It depends on your platform, with Android for example, you can use TelephonyManager

```JAVA
TelephonyManager tm = (TelephonyManager) getSystemService(Context.TELEPHONY_SERVICE);
```
```java
tm.getDeviceId();
```

With the sufficent permissions of course

```xml
<uses-permission android:name="android.permission.READ_PHONE_STATE"/>
```

#### Password Overview

- **Android**: MD5 hash of reversed IMEI (Credit: [WhatsAPI Original Authors](https://github.com/venomous0x/WhatsAPI#venomous))
- **iOS**: MD5 hash of the MAC address repeated twice (Credit: [Ezio Amodio](http://www.ezioamodio.it/?p=29))
- **Windows Phone**: MD5 hash of reversed DeviceUniqueId (Credit: [Robe Fernández](http://blog-robefernandez.blogspot.com.es/2012/11/know-your-whatsapp-password-on-windows.html))

# NOTES

- This proof of concept is extensible to contain every feature that make a fully-fledged client, similar to the official ones, actually could be even better.

- During the two weeks of analysis of service mechanisms, we stumbled upon serious design and security flaws (they fixed some of them since 2011). For a company with such massive user base, we expected better practises and engineering.

- Perfectly working as PHP and JAVA ports.

# License

MIT - refer to the source code for the extra line.

# Venomous

Team of Bahraini Developers.

Ahmed Moh'd ([fb.com/ahmed.mhd](https://www.facebook.com/ahmed.mhd)) and Ali Hubail ([@hubail](https://twitter.com/hubail)) contributed to this release.


CHANGELOG.md
Feb 14, 2013
============
- Merge pull request #271 from facine/master Fix properties typo. && Add auto-upload images to WhatsApp servers and auto-generate icon.
- Critical: Fix properties typo. Update functions doc comment in WhatsProt class. Add auto-message id to: Location, sendStatusUpdate and RequestLastSeen. And add auto-upload images to WhatsApp servers and auto-generate icon.
- Merge pull request #268 from facine/master.
- Add support to send messages to groups.
- Fix property typo.
- Convert code using Sensio Labs' PHP Coding Standars Fixer.

Feb 13, 2013
============
- Merge pull request #267 from facine/master Add support for new registration and identification method.
- Remove partial registration method file.
- Update WhatsApp version.
- Add check response in request and registration code.
- Add support for new registration method.
- Add support for new identification method.
- Remove unused properties.

Feb 04, 2013
============
- Merge pull request #260 from TheKirk/master Merged, hopefully it'll fully integrated into the lib later.

Feb 01, 2013
============
- Added partial support for Registration v2.

Nov 28, 2012
============
- Merge pull request #166 from wertarbyte/fix_RequestLastSeen add missing msgid parameters to RequestLastSeen.
- Updated README.md.
- Merge branch 'beldar-master'.
- Checkpoint-28-11.
- Merge branch 'master' of https://github.com/beldar/WhatsAPI into beldar-master.

Nov 27, 2012
============
- Lots of changes, new outqueue system, auto msgid, auto send nickname and presence, node processing, and more see testprotocol.php for more details.

Nov 26, 2012
============
- Implemented SendPrecense, WaitforReceipt working for sending several images in sequence, and also notification node in messages to see notification on phone on message received.

Nov 25, 2012
============
- Merge pull request #173 from alexandresaiz/patch-1. Adding Whatsappify as a real case using WhatsAPI.
- Merge pull request #184 from meandrex/master. Added commandline argument to set status.

Nov 20, 2012
============
- Added commandline argument to set the Status. Check default help for usage.

Nov 17, 2012
============
- Adding Whatsappify as a real case using WhatsAPI.

Nov 16, 2012
============
- Merge pull request #169 from dleivag/master. New sendPresence method.
- Update src/php/whatsprot.class.php. Fix error.

Nov 15, 2012
============
- Update src/php/whatsprot.class.php. Fix mistake in sendPresence method.
- Update src/php/whatsprot.class.php. Add method sendPresence. It allows to switch between states online - offline.

Nov 14, 2012
============
- Add missing msgid parameters to RequestLastSeen.

Nov 13, 2012
============
- Merge pull request #77 from arthursc/master. Add nickname function for Iphone client.
- Merge pull request #162 from dannydamianus/patch-1. Update src/php/whatsprot.class.php.
- Update src/php/whatsprot.class.php. Added set status support.

Nov 10, 2012
============
- Merge pull request #157 from beldar/master. Fix for sending images.
- Merge pull request #156 from atans/master. Remove php close tag.
- Fix for sending images.

Nov 09, 2012
============
- Update src/php/whatsprot.class.php. Remove php close tag.
- Update src/php/rc4.php. Remove php close tag.
- Update src/php/protocol.class.php. Remove php close tag.
- Update src/php/func.php. Remove php close tag.
- Update src/php/exception.php. Remove php close tag.
- Update src/php/decode.php. Remove php close tag.

Nov 08, 2012
============
- Merge pull request #152 from zjorzzzey/master. Fixes crash while polling for messages.
- Fixes crash when more than one node is sent to BinTreeNodeReader::nextTree().

Nov 07, 2012
============
- Merge pull request #144 from zstars/patch-1. Fixed extremely minor bug in peekInt8, fixed #141.
- Fixed extremely minor bug in peekInt8. https://github.com/venomous0x/WhatsAPI/issues/141.
- Merge pull request #138 from lowlevel-studios/master. Implemented protocol 1.2, fixes #126 #127 #130 #134 #135 #136.
- Implemented protocol 1.2 (php).

Oct 10, 2012
============
- Merge pull request #105 from arthursc/master. Added Location & Requestlastseen.

Oct 06, 2012
============
- Added RequestLastSeen.
- Update src/php/whatsprot.class.php. Added Location function.

Sep 29, 2012
============
- Update README.md.

Sep 18, 2012
============
- Added sendNickname function. A nickname has to be sent at the beginning of the session. When not set the Iphone client will display the phonenumber instead of the name. The nickname can be changed during the session.
- Added contrib notes.

Sep 16, 2012
============
- Merge pull request #66 from arthursc/master. [#64].
- Update src/php/whatsprot.class.php.
- Update src/php/whatsprot.class.php.
- Updated MD files, got rid of unused file. Added Pong function (to respond on Whatsapp's PING request): http://xmpp.org/extensions/xep-0199.html#c2s server to client ping. Added if statement to handle incoming PING requests in processInboundData.
- Merge pull request #56 from chid/master. Added iOS password method.
- Merge pull request #63 from cdesjardins/master. Added ability to send images.
- Fix file path in test code.
- Adding image send.

Sep 15, 2012
============
- Added iOS password method.

Sep 13, 2012
============
- Merge pull request #53 from cdesjardins/master. Had a problem handling incomplete messages.
- Add message to incomplete message exception.

Sep 12, 2012
============
- Merge pull request #47 from cdesjardins/master. Login for iOS Accounts.
- Remote annoying printf, if you want to see debug info use -d.
- Added dummy function RequestLastSeen.
- Quick fixes/typo-fix.
- Merge pull request #49 from Phhere/master.
- Code for iOS added, Accountinfo fixed.
- Login fixed for Android/iOS.
- Fixed Password check.
- Update tests/whatsapp.php. Added password check.
- Update tests/whatsapp.php. Make it runable.
- Merge branch 'master' of git://github.com/venomous0x/WhatsAPI.
- Fix issue 44, can now send special chars.

Sep 11, 2012

php/decode.php
<?php

function decode($hex){
	$hexarr = str_split($hex, 2);
	$str = null;
	foreach($hexarr as $k=>$v)	$str .= "  ".getToken(hexdec($v));
	return $str;
}

function str2hex($string) {
	$hexstr = unpack('H*', $string);
	return array_shift($hexstr);
}

function hex2str($hexstr) {
	$hexstr = str_replace(' ', '', $hexstr);
	$hexstr = str_replace('\x', '', $hexstr);
	$retstr = pack('H*', $hexstr);
	return $retstr;
}

function printhexstr($data, $name)
{
    $data = str2hex($data);
    $len = strlen($data);
    print("Len: $len - $name\n");
    for ($i = 0; $i < $len; $i += 2)
    {
        if ((($i-1) % 32) == 31)
        {
            print("\n");
        }
        printf(" %s%s", $data[$i], $data[$i+1]);
    }

    print("\n");
}

function getDictionary(){
    $dict = "WHATISTHIS";
	$dic[0] = 0;
	$dic[1] = 0;
	$dic[2] = 0;
	$dic[3] = 0;
	$dic[4] = 0;
	$dic[5] = "1";
	$dic[6] = "1.0";
	$dic[7] = "ack";
	$dic[8] = "action";
	$dic[9] = "active";
	$dic[10] = "add";
	$dic[11] = "all";
	$dic[12] = "allow";
	$dic[13] = "apple";
	$dic[14] = "audio";
	$dic[15] = "auth";
	$dic[16] = "author";
	$dic[17] = "available";
	$dic[18] = "bad-request";
	$dic[19] = "basee64";
	$dic[20] = "Bell.caf";
	$dic[21] = "bind";
	$dic[22] = "body";
	$dic[23] = "Boing.caf";
	$dic[24] = "cancel";
	$dic[25] = "category";
	$dic[26] = "challenge";
	$dic[27] = "chat";
	$dic[28] = "clean";
	$dic[29] = "code";
	$dic[30] = "composing";
	$dic[31] = "config";
	$dic[32] = "conflict";
	$dic[33] = "contacts";
	$dic[34] = "create";
	$dic[35] = "creation";
	$dic[36] = "default";
	$dic[37] = "delay";
	$dic[38] = "delete";
	$dic[39] = "delivered";
	$dic[40] = "deny";
	$dic[41] = "DIGEST-MD5";
	$dic[42] = "DIGEST-MD5-1";
	$dic[43] = "dirty";
	$dic[44] = "en";
	$dic[45] = "enable";
	$dic[46] = "encoding";
	$dic[47] = "error";
	$dic[48] = "expiration";
	$dic[49] = "expired";
	$dic[50] = "failure";
	$dic[51] = "false";
	$dic[52] = "favorites";
	$dic[53] = "feature";
	$dic[54] = "field";
	$dic[55] = "free";
	$dic[56] = "from";
	$dic[57] = "g.us";
	$dic[58] = "get";
	$dic[59] = "Glas.caf";
	$dic[60] = "google";
	$dic[61] = "group";
	$dic[62] = "groups";
	$dic[63] = "g_sound";
	$dic[64] = "Harp.caf";
	$dic[65] = "http://etherx.jabber.org/streams";
	$dic[66] = "http://jabber.org/protocol/chatstates";
	$dic[67] = "id";
	$dic[68] = "image";
	$dic[69] = "img";
	$dic[70] = "inactive";
	$dic[71] = "internal-server-error";
	$dic[72] = "iq";
	$dic[73] = "item";
	$dic[74] = "item-not-found";
	$dic[75] = "jabber:client";
	$dic[76] = "jabber:iq:last";
	$dic[77] = "jabber:iq:privacy";
	$dic[78] = "jabber:x:delay";
	$dic[79] = "jabber:x:event";
	$dic[80] = "jid";
	$dic[81] = "jid-malformed";
	$dic[82] = "kind";
	$dic[83] = "leave";
	$dic[84] = "leave-all";
	$dic[85] = "list";
	$dic[86] = "location";
	$dic[87] = "max_groups";
	$dic[88] = "max_participants";
	$dic[89] = "max_subject";
	$dic[90] = "mechanism";
	$dic[91] = "mechanisms";
	$dic[92] = "media";
	$dic[93] = "message";
	$dic[94] = "message_acks";
	$dic[95] = "missing";
	$dic[96] = "modify";
	$dic[97] = "name";
	$dic[98] = "not-acceptable";
	$dic[99] = "not-allowed";
	$dic[100] = "not-authorized";
	$dic[101] = "notify";
	$dic[102] = "Offline Storage";
	$dic[103] = "order";
	$dic[104] = "owner";
	$dic[105] = "owning";
	$dic[106] = "paid";
	$dic[107] = "participant";
	$dic[108] = "participants";
	$dic[109] = "participating";
	$dic[110] = "fail";
	$dic[111] = "paused";
	$dic[112] = "picture";
	$dic[113] = "ping";
	$dic[114] = "PLAIN";
	$dic[115] = "platform";
	$dic[116] = "presence";
	$dic[117] = "preview";
	$dic[118] = "probe";
	$dic[119] = "prop";
	$dic[120] = "props";
	$dic[121] = "p_o";
	$dic[122] = "p_t";
	$dic[123] = "query";
	$dic[124] = "raw";
	$dic[125] = "receipt";
	$dic[126] = "receipt_acks";
	$dic[127] = "received";
	$dic[128] = "relay";
	$dic[129] = "remove";
	$dic[130] = "Replaced by new connection";
	$dic[131] = "request";
	$dic[132] = "resource";
	$dic[133] = "resource-constraint";
	$dic[134] = "response";
	$dic[135] = "result";
	$dic[136] = "retry";
	$dic[137] = "rim";
	$dic[138] = "s.whatsapp.net";
	$dic[139] = "seconds";
	$dic[140] = "server";
	$dic[141] = "session";
	$dic[142] = "set";
	$dic[143] = "show";
	$dic[144] = "sid";
	$dic[145] = "sound";
	$dic[146] = "stamp";
	$dic[147] = "starttls";
	$dic[148] = "status";
	$dic[149] = "stream:error";
	$dic[150] = "stream:features";
	$dic[151] = "subject";
	$dic[152] = "subscribe";
	$dic[153] = "success";
	$dic[154] = "system-shutdown";
	$dic[155] = "s_o";
	$dic[156] = "s_t";
	$dic[157] = "t";
	$dic[158] = "TimePassing.caf";
	$dic[159] = "timestamp";
	$dic[160] = "to";
	$dic[161] = "Tri-tone.caf";
	$dic[162] = "type";
	$dic[163] = "unavailable";
	$dic[164] = "uri";
	$dic[165] = "url";
	$dic[166] = "urn:ietf:params:xml:ns:xmpp-bind";
	$dic[167] = "urn:ietf:params:xml:ns:xmpp-sasl";
	$dic[168] = "urn:ietf:params:xml:ns:xmpp-session";
	$dic[169] = "urn:ietf:params:xml:ns:xmpp-stanzas";
	$dic[170] = "urn:ietf:params:xml:ns:xmpp-streams";
	$dic[171] = "urn:xmpp:delay";
	$dic[172] = "urn:xmpp:ping";
	$dic[173] = "urn:xmpp:receipts";
	$dic[174] = "urn:xmpp:whatsapp";
	$dic[175] = "urn:xmpp:whatsapp:dirty";
	$dic[176] = "urn:xmpp:whatsapp:mms";
	$dic[177] = "urn:xmpp:whatsapp:push";
	$dic[178] = "value";
	$dic[179] = "vcard";
	$dic[180] = "version";
	$dic[181] = "video";
	$dic[182] = "w";
	$dic[183] = "w:g";
	$dic[184] = "w:p:r";
	$dic[185] = "wait";
	$dic[186] = "x";
	$dic[187] = "xml-not-well-formed";
	$dic[188] = "xml:lang";
	$dic[189] = "xmlns";
	$dic[190] = "xmlns:stream";
	$dic[191] = "Xylophone.caf";
	$dic[192] = "account";
	$dic[193] = "digest";
	$dic[194] = "g_notify";
	$dic[195] = "method";
	$dic[196] = "password";
	$dic[197] = "registration";
	$dic[198] = "stat";
	$dic[199] = "text";
	$dic[200] = "user";
	$dic[201] = "username";
	$dic[202] = "event";
	$dic[203] = "latitude";
	$dic[204] = "longitude";
	$dic[205] = "true";
	$dic[206] = "after";
	$dic[207] = "before";
	$dic[208] = "broadcast";
	$dic[209] = "count";
	$dic[210] = "features";
	$dic[211] = "first";
	$dic[212] = "index";
	$dic[213] = "invalid-mechanism";
	$dic[214] = "l$dict";
	$dic[215] = "max";
	$dic[216] = "offline";
	$dic[217] = "proceed";
	$dic[218] = "required";
	$dic[219] = "sync";
	$dic[220] = "elapsed";
	$dic[221] = "ip";
	$dic[222] = "microsoft";
	$dic[223] = "mute";
	$dic[224] = "nokia";
	$dic[225] = "off";
	$dic[226] = "pin";
	$dic[227] = "pop_mean_time";
	$dic[228] = "pop_plus_minus";
	$dic[229] = "port";
	$dic[230] = "reason";
	$dic[231] = "server-error";
	$dic[232] = "silent";
	$dic[233] = "timeout";
	$dic[234] = "lc";
	$dic[235] = "lg";
	$dic[236] = "bad-protocol";
	$dic[237] = "none";
	$dic[238] = "remote-server-timeout";
	$dic[239] = "service-unavailable";
	$dic[240] = "w:p";
	$dic[241] = "w:profile:picture";
	$dic[242] = "notification";
	$dic[243] = 0;
	$dic[244] = 0;
	$dic[245] = 0;
	$dic[246] = 0;
	$dic[247] = 0;
    $dic[248] = "XXX";
    return $dic;
}

function getToken($token){
    $dic = getDictionary();
	return $dic[$token];
}

		
?>
